sim/battle-queue.ts

```diff
@@ -0,0 +1,367 @@
+	/** fractional priority of the action (lower first) */
+	fractionalPriority: number;
+	/** speed of pokemon using move (higher first if priority tie) */
+	speed: number;
+	/** the pokemon doing the move */
+	pokemon: Pokemon;
+	/** location of the target, relative to pokemon's side */
+	targetLoc: number;
+	/** original target pokemon, for target-tracking moves */
+	originalTarget: Pokemon;
+	/** a move to use (move action only) */
+	moveid: ID;
+	/** a move to use (move action only) */
+	move: Move;
+	/** true if megaing or ultra bursting */
+	mega: boolean | 'done';
+	/** if zmoving, the name of the zmove */
+	zmove?: string;
+	/** if dynamaxed, the name of the max move */
+	maxMove?: string;
+	/** effect that called the move (eg Instruct) if any */
+	sourceEffect?: Effect | null;
+}
+
+/** A switch action */
+export interface SwitchAction {
+	/** action type */
+	choice: 'switch' | 'instaswitch';
+	order: 3 | 103;
+	/** priority of the action (lower first) */
+	priority: number;
+	/** speed of pokemon switching (higher first if priority tie) */
+	speed: number;
+	/** the pokemon doing the switch */
+	pokemon: Pokemon;
+	/** pokemon to switch to */
+	target: Pokemon;
+	/** effect that called the switch (eg U */
+	sourceEffect: Effect | null;
+}
+
+/** A Team Preview choice action */
+export interface TeamAction {
+	/** action type */
+	choice: 'team';
+	/** priority of the action (lower first) */
+	priority: number;
+	/** unused for this action type */
+	speed: 1;
+	/** the pokemon switching */
+	pokemon: Pokemon;
+	/** new index */
+	index: number;
+}
+
+/** A generic action not done by a pokemon */
+export interface FieldAction {
+	/** action type */
+	choice: 'start' | 'residual' | 'pass' | 'beforeTurn';
+	/** priority of the action (lower first) */
+	priority: number;
+	/** unused for this action type */
+	speed: 1;
+	/** unused for this action type */
+	pokemon: null;
+}
+
+/** A generic action done by a single pokemon */
+export interface PokemonAction {
+	/** action type */
+	choice: 'megaEvo' | 'shift' | 'runPrimal' | 'runSwitch' | 'event' | 'runUnnerve' | 'runDynamax';
+	/** priority of the action (lower first) */
+	priority: number;
+	/** speed of pokemon doing action (higher first if priority tie) */
+	speed: number;
+	/** the pokemon doing action */
+	pokemon: Pokemon;
+}
+
+export type Action = MoveAction | SwitchAction | TeamAction | FieldAction | PokemonAction;
+
+/**
+ * An ActionChoice is like an Action and has the same structure, but it doesn't need to be fully filled out.
+ *
+ * Any Action or ChosenAction qualifies as an ActionChoice.
+ *
+ * The `[k: string]: any` part is required so TypeScript won't warn about unnecessary properties.
+ */
+export interface ActionChoice {
+	choice: string;
+	[k: string]: any;
+}
+
+/**
+ * Kind of like a priority queue, although not sorted mid-turn in Gen 1-7.
+ *
+ * Sort order is documented in `BattleQueue.comparePriority`.
+ */
+export class BattleQueue extends Array<Action> {
```

Comment:
I'm still not super convinced about `extends Array` instead of having an array as a property (which among other things means we can't overwrite the queue with `= []`.

Certain awkwardnesses:

`queue = []` becomes `queue.splice(0)`

`queue = array` becomes `queue.splice(0); queue.push(...array)`

`queue.slice()` becomes `[...queue]`

`queue = queue.filter` becomes "rewrite it as a for-loop"

Still, though, it's kinda slick, and I don't mind the workarounds. It's easy enough to back out of. In theory.

Reply:
(I'm merging now, but please do still give feedback on this.)